### PR TITLE
Refactor: simpler read

### PR DESCRIPTION
**Pull request by gnuser**

### CODE REVIEW

File: src/streams.h [hunk 1]
```diff
@@ -345,18 +345,16 @@ class CDataStream
 
         // Read from the beginning of the buffer
         unsigned int nReadPosNext = nReadPos + nSize;
-        if (nReadPosNext >= vch.size())
+        if (nReadPosNext > vch.size()) {
+            throw std::ios_base::failure("CDataStream::read(): end of data");
+        }
+        memcpy(pch, &vch[nReadPos], nSize);
```
Comment (sipa) [inline src/streams.h]: This is not correct. If nReadPos = 0, and nSize = 0, this will call `operator[]` on `vch`, which is illegal when it is empty.

Comment (MeshCollider) [reply src/streams.h]: Note that that's a bug with the original code though, not with the refactor

**Review by dcousens — APPROVED** [review]

File: src/streams.h (hunk 1 above)
Comment (laanwj) [reply src/streams.h]: probably better to use `vch.data() + nNreadPos`?

Comment (promag) [reply src/streams.h]: @sipa above there is an early return if `nSize == 0` so that doesn't happen because of the `throw`.

**Comment by promag** [comment]
utACK 9db9d62.

Nit, commit message could mention `CDataStream::read` instead `read` only.

File: src/streams.h [hunk 2]
```diff
@@ -345,18 +345,16 @@ class CDataStream
 
         // Read from the beginning of the buffer
         unsigned int nReadPosNext = nReadPos + nSize;
-        if (nReadPosNext >= vch.size())
+        if (nReadPosNext > vch.size()) {
+            throw std::ios_base::failure("CDataStream::read(): end of data");
+        }
+        memcpy(pch, &vch[nReadPos], nSize);        
+        if (nReadPosNext == vch.size())
```
Comment (promag) [inline src/streams.h]: Nit, `{` on this line.

File: src/streams.h (hunk 1 above)
Comment (sipa) [reply src/streams.h]: Agree, there is no bug here. It would still be clearer using `vch.data() + nReadPos` though.

**Pull request merged**